src/Collateral.sol

```diff
@@ -0,0 +1,202 @@
+
+        emit ThawRequest(msg.sender, receiver, amount, account.thawEnd);
+    }
+
+    /**
+     * @dev Withdraws all thawed collateral from a receivers collateral account.
+     * @param receiver Address of the receiver.
+     * @notice REVERT: this function will revert if the sender receiver collateral account does
+     *                not have any thawed collateral. This function will also revert if no thawing
+     *               period has been completed.
+     */
+    function withdrawThawedCollateral(address receiver) external {
+        CollateralAccount storage account = collateralAccounts[msg.sender][receiver];
+        require(account.thawEnd != 0, "No collateral thawing");
+        require(account.thawEnd <= block.number, "Collateral still thawing");
+
+        // Amount is the minimum between the amount being thawed and the actual balance
+        uint256 amount = account.amountThawing > account.balance ? account.balance : account.amountThawing;
+
+        unchecked {
+            account.balance -= amount; // Reduce the balance by the withdrawn amount (no underflow risk)
+        }
+        account.amountThawing = 0;
+        account.thawEnd = 0;
+        emit Withdraw(msg.sender, receiver, amount);
+        require(collateralToken.transfer(msg.sender, amount));
+    }
+
+    /**
+     * @dev Authorizes a signer to sign RAVs for the sender.
+     * @param signer Address of the authorized signer.
+     */
+    function authorizeSigner(address signer) external {
+        require(authorizedSigners[signer] == address(0), "Signer already authorized");
+        authorizedSigners[signer] = msg.sender;
+        emit AuthorizeSigner(signer, msg.sender);
+    }
+
+    /**
+     * @dev Redeems collateral for a receiver using a signed RAV.
+     * @param signedRAV Signed RAV containing the receiver and collateral amount.
+     * @notice REVERT: this function will revert if:
+     *                  - the signer is not authorized to sign for a sender.
+     *                  - the sender receiver collateral account does not have enough
+     *                    collateral (greater than the amount in the RAV).
+     *                  - the allocation ID has already been used.
+     */
+    function redeem(TAPVerifier.SignedRAV memory signedRAV) external {
+        address signer = tapVerifier.recoverRAVSigner(signedRAV);
+        require(authorizedSigners[signer] != address(0), "Signer not authorized");
+
+        address sender = authorizedSigners[signer];
+        address receiver = msg.sender;
+        uint256 amount = signedRAV.rav.valueAggregate;
+        require(collateralAccounts[sender][receiver].balance >= amount, "Insufficient collateral balance");
+        unchecked {
+            collateralAccounts[sender][receiver].balance -= amount;
+        }
+
+        emit Redeem(msg.sender, amount);
+        allocationIDTracker.useAllocationID(signedRAV.rav.allocationId);
+        require(collateralToken.transfer(msg.sender, amount));
```

Comment:
This design bypasses query rebate system. `Collateral.redeem()` should call `Staking.collect()` instead of directly transferring the GRT tokens to the indexer.

Reply:
We are tracking this one in #6